heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -40,11 +40,11 @@
  */
 public abstract class WakeableLooper {
   // The tasks could only be added but not removed
-  private final List<Runnable> tasksOnWakeup;
+  private final CopyOnWriteArrayList<Runnable> tasksOnWakeup;
```

Comment:
WakeableLooper is designed not to be thread-safe; concurrent scenarios should be handled in outer logic.

If the purpose of this pull request is to support thread-safety, using this concurrent structure is not safe enough, if thinking this from an application point of view, since the actual `Runnable` a removed reference pointing to might not run normally, for instance, running with a dirty state.

If the purpose is to avoid task-removal during execution, I would suggest not to use a `Concurrent` data structure to avoid confusion.

BTW, IIRC, previously we could only add a task to `tasksOnWakeup` when executing tasks from it, but we could not remove a task.

Reply:
Also, if the purpose is to avoid task-removal during execution, why not find out such code snippets and correct them?

Reply:
@maosongfu The code snippet of the task that clears the tasksOnWakeup list:

https://github.com/twitter/heron/blob/master/heron/instance/src/java/com/twitter/heron/instance/Slave.java#L113


Has to do with restoring state.